plugins/apps/threatbus_misp/threatbus_misp/message_mapping.py

```diff
@@ -106,69 +204,122 @@ def is_whitelisted(misp_msg: dict, filter_config: List[Dict]):
+    @return a valid uuid
+    """
+    return stix2_indicator_uuid[len("indicator--") :]
+
+
+def attribute_to_stix2_indicator(
+    misp_attribute: dict, action: str, logger
+) -> Union[None, Update, Indicator]:
+    """
+    Maps the given MISP attribute to a STIX-2 cyber observable object.
+    @param misp_attribute The MISP attribute to map to STIX-2
     @param action A string from MISP, describing the action for the attribute
-        (either 'add' or 'delete')
-    @return The mapped intel item or None
+        (either 'add', 'edit', or 'delete')
+    @param logger The logger instance from the calling function
+    @return The mapped STIX-2 Indicator, a Threat Bus Update, or None
     """
     # parse MISP attribute
     if not misp_attribute:
         return None
     to_ids = misp_attribute.get("to_ids", False)
     if not to_ids and action != "edit" or not action:
         return None
-    operation = Operation.REMOVE
-    if (action == "edit" or action == "add") and to_ids:
-        operation = Operation.ADD
-
-    # parse values
-    intel_type = misp_attribute.get("type", None)
-    indicator = misp_attribute.get("value", None)
-    if not intel_type or not indicator:
+    stix2_id = stix2_indicator_id(misp_attribute["uuid"])
+    stix2_timestamp = datetime.fromtimestamp(int(misp_attribute["timestamp"]))
+    if action == "edit" and not to_ids:
+        return Update(id=stix2_id, operation=Operation.REMOVE)
+
+    ## parse MISP values
+    attr_type = misp_attribute.get("type", None)
+    attr_value = misp_attribute.get("value", None)
+    if not attr_type or not attr_value:
+        logger.debug(
+            f"Incomplete MISP attribute, missing `type` and/or `value` fields: '{misp_attribute}'"
+        )
         return None
 
-    # parse compound MISP intel:
-    if "|" in intel_type:
-        indicator = tuple(indicator.split("|"))
-        if len(indicator) != 2:
-            if logger:
+    ## parse compound MISP intel:
+    if "|" in attr_type:
+        obs_exprs = []  # list to hold all ObservationExpressions
+        value_splits = attr_value.split("|")
+        for idx, attr_type_split in enumerate(attr_type.split("|")):
+            stix2_create_func = attribute_type_map.get(attr_type_split, None)
+            if not stix2_create_func:
                 logger.debug(
-                    f"Expected '|'-separated composite values for MISP intel type {intel_type}"
+                    f"Cannot find matching STIX-2 type for MISP attribute type '{attr_type}'"
                 )
+                return None
+            try:
+                eq_expr = stix2_create_func(
+                    value_splits[idx]
+                )  # EqualitiComparisonExpression
+                obs_expr = ObservationExpression(eq_expr)
+                obs_exprs.append(obs_expr)
+            except Exception as e:
+                logger.error(f"Error creating STIX-2 expression: {e}")
```

Comment:
This try-except seems dangerous, the resulting `obs_exprs` list can become incomplete and the generated `Indicator` would not be equivalent to the MISP attribute. This might be OK depending on the semantics of `|`, which I don't know. If it is OK, it needs an explanatory comment.